docs/source/generating/testing.rst

```diff
@@ -210,103 +221,37 @@ Or we can simplify further by using ``assertHasMessage`` and ``assertHasAction``
-The messages will still be logged normally but you will be able to limit tests to only looking at those messages.
-
-You can log messages to a specific ``Logger``:
+Custom testing setup
+--------------------
 
-.. code-block:: python
-
-    from eliot import Message, Logger
-
-    class YourClass(object):
-        logger = Logger()
-
-        def run(self):
-            # Create a message with two fields, "key" and "value":
-            msg = Message.new(key=123, value=u"hello")
-            # Write the message:
-            msg.write(self.logger)
-
-As well as actions:
+In some cases ``@capture_logging`` may not do what you want.
+You can achieve the same effect, but with more control, with some lower-level APIs:
 
 .. code-block:: python
 
-     from eliot import start_action
-
-     logger = Logger()
-
-     with start_action(logger, action_type=u"store_data"):
-         x = get_data()
-         store_data(x)
-
-Or actions created from ``ActionType``:
-
-.. code-block:: python
-
-    from eliot import Logger
-
-      from myapp.logtypes import LOG_USER_REGISTRATION
-
-      class UserRegistration(object):
-
-          logger = Logger()
-
-          def __init__(self):
-              self.db = {}
-
-          def register(self, username, password, age):
-              self.db[username] = (password, age)
-              msg = LOG_USER_REGISTRATION(
-                   username=username, password=password, age=age)
-              # Notice use of specific logger:
-              msg.write(self.logger)
-
-The tests would then need to do two things:
-
-1. Decorate your test with ``validate_logging`` instead of ``capture_logging``.
-2. Override the logger used by the logging code to use the one passed in to the test.
-
-For example:
-
-.. code-block:: python
-
-    from eliot.testing import LoggedMessage, validate_logging
-
-    class LoggingTests(TestCase):
-        def assertRegistrationLogging(self, logger):
-            """
-            Logging assertions for test_registration.
-            """
-            logged = LoggedMessage.of_type(logger.messages, LOG_USER_REGISTRATION)[0]
-            assertContainsFields(self, logged.message,
-                                 {u"username": u"john",
-                                  u"password": u"password",
-                                  u"age": 12}))
-
-        # validate_logging only captures log messages logged to the MemoryLogger
-        # instance it passes to the test:
-        @validate_logging(assertRegistrationLogging)
-        def test_registration(self, logger):
-            """
-            Registration adds entries to the in-memory database.
-            """
-            registry = UserRegistration()
-            # Override logger with one used by test:
-            registry.logger = logger
-            registry.register(u"john", u"password", 12)
-            self.assertEqual(registry.db[u"john"], (u"password", 12))
+   from eliot import MemoryLogger
```

Comment:
@exarkun this is the proposed API:

Reply:
`swap_logger` looks alright.  I think this is kind of a standard pattern for managing global singletons.  I have some bad feelings when I look at this example but I think they all stem from the global singleton nature of the resource being managed.  So I guess I can ignore them.

It looks like `check_for_errors` raises so the two calls in the `finally` probably want to happen in the other order so that the original logger is actually swapped back in.  And maybe there's room for a slightly higher-level API that wraps this call in a context manager so it's easy to safely temporarily swap in a different logger.  Not sure that's an API I would use because Deferreds ... but maybe it is.  Depends on exactly how Eliot/test runner integration ends up working.

For what it's worth, this is what I've been using since filing this issue (testtools flavored):  https://github.com/tahoe-lafs/tahoe-lafs/blob/1432.watchdog-magic-folder-with-eliot/src/allmydata/test/eliotutil.py#L115-L146